### PR TITLE
Add more clarity to LRO guidelines regarding exception behavior

### PR DESCRIPTION
**Pull request by christothes**
# Summary
The focus of this proposed change is to provide more clarity to the behavior of the `Value` property of `Operation<T>`, specifically with regard to exceptions.

This guidance is currently implied only in the code LRO code example and the exception behavior comes up frequently in code reviews.

### CODE REVIEW

File: docs/dotnet/introduction.md [hunk 1]
```diff
@@ -505,6 +505,12 @@ BlobBaseClient client = ...
 
 {% include requirement/MUST id="dotnet-lro-return" %} return a subclass of ```Operation<T>``` from LRO methods.
 
+{% include requirement/MUST id="dotnet-lro-return" %} throw ```InvalidOperationException``` when the ```Value``` property is evaluated before the operation is complete (```HasCompleted``` is false).
+
+{% include requirement/MUSTNOT id="dotnet-namespaces-location" %} throw from an LRO except for when evaluating the ```Value``` property as described above.
```
Comment (maririos) [inline docs/dotnet/introduction.md:510]: could you clarify what do you mean with `throw form an LRO` ?

Comment (christothes) [reply docs/dotnet/introduction.md:510]: Primarily, calls  `UpdateStatus` and `WaitForCompletion` should not throw.

File: docs/dotnet/introduction.md [hunk 2]
```diff
@@ -505,6 +505,12 @@ BlobBaseClient client = ...
 
 {% include requirement/MUST id="dotnet-lro-return" %} return a subclass of ```Operation<T>``` from LRO methods.
 
+{% include requirement/MUST id="dotnet-lro-return" %} throw ```InvalidOperationException``` when the ```Value``` property is evaluated before the operation is complete (```HasCompleted``` is false).
```
Comment (annelo-msft) [inline docs/dotnet/introduction.md:508]: I'd suggest putting the "Only throw from the `Value` property" first, then give us the two options for throwing from `Value` next.

File: docs/dotnet/introduction.md (hunk 1 above)
Comment (annelo-msft) [reply docs/dotnet/introduction.md:510]: Maybe say don't throw from methods or properties of `Operation\<T\>` subclasses other than the `Value` property?

File: docs/dotnet/introduction.md [hunk 3]
```diff
@@ -505,6 +505,12 @@ BlobBaseClient client = ...
...
+
+{% include requirement/MUSTNOT id="dotnet-namespaces-location" %} throw from an LRO except for when evaluating the ```Value``` property as described above.
+
+{% include requirement/MUST id="dotnet-lro-return" %} throw ```RequestFailedException``` or its subtype when the ```Value``` property is evaluated and the operation has completed without a successful result.
+
 {% include requirement/MAY id="dotnet-lro-subclass" %} add additional APIs to subclasses of ```Operation<T>```.
```
Comment (annelo-msft) [inline docs/dotnet/introduction.md:514]: Should this guideline stay close to the other "subclass of `Operation\<T\>`" guideline?  That way the logical progression is rougly ...
- Here is an LRO
- Here's how to name LRO methods on clients
- Here's how to design LRO method return types (i.e. `Operation\<T\>` subclasses)
- Here's how to throw an exception from `Operation\<T\>` subclasses
  - `InvalidOperationException`
  - `RequestFailedException`

File: docs/dotnet/introduction.md (hunk 1 above)
Comment (maririos) [reply docs/dotnet/introduction.md]: Hmm we do this in FR. Could you expand on why not (here, not guideline)?

File: docs/dotnet/introduction.md [hunk 4]
```diff
@@ -505,6 +505,12 @@ BlobBaseClient client = ...
 
 {% include requirement/MUST id="dotnet-lro-return" %} return a subclass of ```Operation<T>``` from LRO methods.
 
+{% include requirement/MUST id="dotnet-lro-return" %} throw ```InvalidOperationException``` when the ```Value``` property is evaluated before the operation is complete (```HasCompleted``` is false).
+
+{% include requirement/MUSTNOT id="dotnet-namespaces-location" %} throw from ```UpdateStatus```, ```WaitForCompletion```, or ```WaitForCompletionAsync```.
```
Comment (annelo-msft) [inline docs/dotnet/introduction.md:510]: Could this be "Do not throw an exception from `Operation\<T\>` members other than `Value` ?

Comment (christothes) [reply docs/dotnet/introduction.md:510]: After considering your other comment to change the throw guidance to `Only throw ...` I'm thinking that this additional bullet may be unnecessary. The **only** clarification may be sufficient. Thoughts?

File: docs/dotnet/introduction.md (hunk 1 above)
Comment (christothes) [reply docs/dotnet/introduction.md]: In terms of being predictable on when an LRO can throw, if we say that evaluation of `Value` can and should throw under some circumstances, it would seem unexpected for it to throw just before completing.

Comment (christothes) [reply docs/dotnet/introduction.md]: > Maybe say don't throw exceptions from methods or properties of `Operation\<T\>` subclasses other than the `Value` property?
> 
> What if `UpdateStatus` threw an `OutOfMemoryException`. Would that be valid?

Good question. I think a good clarification here would be that we shouldn't prevent any exception from being thrown (such as an unexpected exception like `OutOfMemopry`, but we shouldn't intentionally throw or allow a `RequestFailedException` to bubble up.

Comment (maririos) [reply docs/dotnet/introduction.md]: So in FR we do:
https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/formrecognizer/Azure.AI.FormRecognizer/tests/samples/Sample1_RecognizeContentFromFile.cs#L28
 
What u are proposing means we wouldn't be able to do that convenience for the user and we should do 
```
var operation = await client.StartRecognizeContentAsync(stream);
operation.WaitForCompletionAsync()
try
{
  FormPageCollection formPages = operation.Value
}
...
```
or something similar.
 
What if the user never calls `operation.Value?` 
i.e.
```
var operation = await client.StartRecognizeContentAsync(stream);
FormPageCollection formPages = operation.WaitForCompletionAsync()
```

Comment (maririos) [reply docs/dotnet/introduction.md]: So we throw on `Value` and `WaitForCompletionAsync`, `UpdateStatus` so the behavior is consistent, no matter where the user is calling the LRO from.

Comment (christothes) [reply docs/dotnet/introduction.md]: Great question! One could argue that success is implied below:
```c#
var operation = await client.StartRecognizeContentAsync(stream);
FormPageCollection formPages = operation.WaitForCompletionAsync()

Console.WriteLine("the operation must have succeeded if you can see this");
```

This seems equivalent to if we didn't throw on failed service operations and instead had to inspect the Response Status to see if it succeeded.

Maybe we should change the guidance to the opposite of what was originally proposed to: 
- Only throw intentionally in `UpdateStatus` or `WaitForCompletion` if the operation fails.

Although it would then seem valid to throw at least `InvalidOperationException` from `Value` if we've already thrown from the operation methods or the operation is not complete.

I've been thinking that a `TryGetValue` would also address the issue of avoiding unexpected throws from accessing `Value`.

For example, I could see this exception being unexpected by a developer customer:

```c#
var operation = await client.StartRecognizeContentAsync(stream);
FormPageCollection formPages = operation.WaitForCompletionAsync()

Console.WriteLine("the operation must have succeeded if you can see this");

// This is definitly safe!
var result = operation.Value;
```

Comment (maririos) [reply docs/dotnet/introduction.md]: I see your point. Part of me (expressed in our [FR tests](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/formrecognizer/Azure.AI.FormRecognizer/tests/FormRecognizerClient/FormRecognizerClientLiveTests.cs#L1735)) was assuming that people will check `HasValue` before consuming `Value`, specially if looking at guarding their code from `Value` throwing, which of course, is not true :).

How will adding a `TryGetValue` help here? Follows the pattern we have in other types where if `Value` throws, it is acceptable because of the existance of `TyGetValue`?

As for the exception from `Value`, in FR we decided to throw the same exception that you get from the LRO method to mantain consistency.

Comment (christothes) [reply docs/dotnet/introduction.md]: > How will adding a `TryGetValue` help here? Follows the pattern we have in other types where if `Value` throws, it is acceptable because of the existance of `TyGetValue`?

I think the existence of `TryGetValue` could be a more of a hint that `Value` can throw. 

> As for the exception from `Value`, in FR we decided to throw the same exception that you get from the LRO method to maintain consistency.

That sounds reasonable, but it feels like `InvalidOperationException` is more correct if `Value` was evaluated after we already clearly failed. This is another reason why I think `TryGetValue` seems better than `Value`. Having the `Value` property available to evaluate and throw seems wrong when we know the operation failed.

File: docs/dotnet/introduction.md (hunk 4 above)
Comment (pakrym) [reply docs/dotnet/introduction.md:510]: Wouldn't `WaitForCompletion` throw if `Value` throws?

Comment (christothes) [reply docs/dotnet/introduction.md:510]: I believe so. Check out the discussion on [this comment](https://github.com/Azure/azure-sdk/pull/2003#discussion_r520145628)

File: docs/dotnet/introduction.md (hunk 2 above)
Comment (tg-msft) [reply docs/dotnet/introduction.md:508]: Can you include the message too?  It'd be great if we all throw the same message for this the same way we (try to) throw the same message for invalid versions.

File: docs/dotnet/introduction.md [hunk 5]
```diff
@@ -505,6 +505,12 @@ BlobBaseClient client = ...
...
+
+{% include requirement/MUST id="dotnet-lro-return" %} throw ```RequestFailedException``` or its subtype when the ```Value``` property is evaluated and the operation has completed without a successful result.
```
Comment (tg-msft) [inline docs/dotnet/introduction.md:512]: Consider adding something about trying to short circuit out of `UpdateStatus` if `HasCompleted` like https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/storage/Azure.Storage.Blobs/src/Models/CopyFromUriOperation.cs#L157 to avoid exceptions being thrown if you're trying to mock an operation?

File: docs/dotnet/introduction.md [hunk 6]
```diff
@@ -505,9 +505,22 @@ BlobBaseClient client = ...
...
 
+{% include requirement/MUST id="dotnet-lro-return" %} throw from ```Operation<T>``` subclasses in the following scenarios.
+
+* ```RequestFailedException``` or its subtype should be thrown from ```UpdateStatus```, ```WaitForCompletion```, or ```WaitForCompletionAsync``` if the operation completes with an non-success result. Include any relevant error state information in the exception details.
+* Any ```RequestFailedException```or its subtype resulting from the underlying service operation calls in ```RequestFailedException``` from ```UpdateStatus```, ```WaitForCompletion```, or ```WaitForCompletionAsync```.
+* ```RequestFailedException``` should be thrown when the ```Value``` property is evaluated before the operation is complete (```HasCompleted``` is false).
+  * The exception should be the same as what was thrown from ```UpdateStatus```, ```WaitForCompletion```, or ```WaitForCompletionAsync```.
+* ```InvalidOperationException``` should be thrown when the ```Value``` property is evaluated if the operation failed (```HasValue``` is false).
+  * The exception message should be: "The operation has not yet completed."
```
Comment (maririos) [inline docs/dotnet/introduction.md:520]: I organized it a little bit to make it more clear. Feel free to adapt as you consider
```suggestion
{% include requirement/MUST id="dotnet-lro-return" %} throw from ```Operation<T>``` subclasses in the following scenarios:

* If the operation completes with an non-success result, throw ```RequestFailedException``` or its subtype from ```UpdateStatus```, ```WaitForCompletion```, or ```WaitForCompletionAsync```. Include any relevant error state information in the exception details.
* If the underlying service operation calls from ```UpdateStatus```, ```WaitForCompletion```, or ```WaitForCompletionAsync``` returns any ```RequestFailedException```or its subtype. Return same exception.
* If the ```Value``` property is evaluated before the operation is complete (```HasCompleted``` is false), throw ```RequestFailedException```.
  * The exception should be the same as what was thrown from ```UpdateStatus```, ```WaitForCompletion```, or ```WaitForCompletionAsync```.
* If the ```Value``` property is evaluated and the operation failed (```HasValue``` is false), ```InvalidOperationException``` should be thrown. 
  * The exception message should be: "The operation has not yet completed."
```
```

File: docs/dotnet/introduction.md [hunk 7]
```diff
@@ -505,9 +505,22 @@ BlobBaseClient client = ...
...
+
 {% include requirement/MAY id="dotnet-lro-subclass" %} add additional APIs to subclasses of ```Operation<T>```.
 For example, some subclasses add a constructor allowing to create an operation instance from a previously saved operation ID. Also, some subclasses are more granular states besides the IsCompleted and HasValue states that are present on the base class.
 
+{% include requirement/MUST id="dotnet-lro-return" %} throw from ```Operation<T>``` subclasses in the following scenarios.
+
+* ```RequestFailedException``` or its subtype should be thrown from ```UpdateStatus```, ```WaitForCompletion```, or ```WaitForCompletionAsync``` if the operation completes with an non-success result. Include any relevant error state information in the exception details.
+* Any ```RequestFailedException```or its subtype resulting from the underlying service operation calls in ```RequestFailedException``` from ```UpdateStatus```, ```WaitForCompletion```, or ```WaitForCompletionAsync```.
```
Comment (maririos) [inline docs/dotnet/introduction.md:516]: I organized it a little bit to make it more clear. Feel free to adapt as you consider
```suggestion
{% include requirement/MUST id="dotnet-lro-return" %} throw from ```Operation<T>``` subclasses in the following scenarios:

* If the operation completes with an non-success result, throw ```RequestFailedException``` or its subtype from ```UpdateStatus```, ```WaitForCompletion```, or ```WaitForCompletionAsync```. Include any relevant error state information in the exception details.
* If the underlying service operation calls from ```UpdateStatus```, ```WaitForCompletion```, or ```WaitForCompletionAsync``` returns any ```RequestFailedException```or its subtype. Return same exception.

File: docs/dotnet/introduction.md (hunk 6 above)
Comment (maririos) [inline docs/dotnet/introduction.md:520]: hmmm should this be flipped?
```
if (HasCompleted == false)
  throw InvalidOperationException
if (HasValue == false) -> also, could this be `if (HasValue == false && HasCompleted == true)`
  throw RequestFailedException
```

File: docs/dotnet/introduction.md (hunk 7 above)
Comment (maririos) [reply docs/dotnet/introduction.md:516]: and write the other ones like this too :)

File: docs/dotnet/introduction.md (hunk 6 above)
Comment (christothes) [reply docs/dotnet/introduction.md:520]: Yes!

File: docs/dotnet/introduction.md [hunk 8]
```diff
@@ -505,9 +505,20 @@ BlobBaseClient client = ...
...
+{% include requirement/MUST id="dotnet-lro-return" %} check the value of ```HasCompleted``` in subclass implementations of ```UpdateStatus``` and ```UpdateStatusAsync``` and immediately return the result of ```GetRawResponse``` if it is true.
+
 {% include requirement/MAY id="dotnet-lro-subclass" %} add additional APIs to subclasses of ```Operation<T>```.
 For example, some subclasses add a constructor allowing to create an operation instance from a previously saved operation ID. Also, some subclasses are more granular states besides the IsCompleted and HasValue states that are present on the base class.
 
+{% include requirement/MUST id="dotnet-lro-return" %} throw from ```Operation<T>``` subclasses in the following scenarios.
+
+* If the operation completes with a non-success result, throw ```RequestFailedException``` or its subtype from ```UpdateStatus```, ```WaitForCompletion```, or ```WaitForCompletionAsync```.
+  * Include any relevant error state information in the exception details.
+* If an underlying service operation call from ```UpdateStatus```, ```WaitForCompletion```, or ```WaitForCompletionAsync``` throws, re-throw ```RequestFailedException``` or its subtype.
```
Comment (annelo-msft) [inline docs/dotnet/introduction.md:517]: Does this conflict with @heaths 's performance best practice:
"For example, don't throw ex in a catch handler where you just captured ex (you rewrite the stack)."

Comment (christothes) [reply docs/dotnet/introduction.md:517]: It is not intended to. Perhaps it needs some better wording. 

What this is intended to say is if the underlying method threw `RequestFailedException` or its sub-type, just `throw`. I added the wording about re-throwing because I was thinking that we may want to wrap the underlying exception of another type into a `RequestFailedException`, like we do my default with service method implementations.

File: docs/dotnet/introduction.md [hunk 9]
```diff
@@ -505,9 +505,20 @@ BlobBaseClient client = ...
 
 {% include requirement/MUST id="dotnet-lro-return" %} return a subclass of ```Operation<T>``` from LRO methods.
 
+{% include requirement/MUST id="dotnet-lro-return" %} check the value of ```HasCompleted``` in subclass implementations of ```UpdateStatus``` and ```UpdateStatusAsync``` and immediately return the result of ```GetRawResponse``` if it is true.
+
```
Comment (KrzysztofCwalina) [inline docs/dotnet/introduction.md:509]: Can you add a code sample illustrating this? Also, I worry that this is an implementation detail not API design guideline. This document is meant to talk about APIs. We have a separate document for implementation guidelines. Alternatively, such implementation details could be explained as commentary below existing related API guidelines. In other words, I don't think this document should have top level bullets/guidelines that talk about implementation details.

File: docs/dotnet/introduction.md [hunk 10]
```diff
@@ -505,9 +505,20 @@ BlobBaseClient client = ...
...
 {% include requirement/MAY id="dotnet-lro-subclass" %} add additional APIs to subclasses of ```Operation<T>```.
 For example, some subclasses add a constructor allowing to create an operation instance from a previously saved operation ID. Also, some subclasses are more granular states besides the IsCompleted and HasValue states that are present on the base class.
 
+{% include requirement/MUST id="dotnet-lro-return" %} throw from ```Operation<T>``` subclasses in the following scenarios.
```
Comment (KrzysztofCwalina) [inline docs/dotnet/introduction.md:513]: Types don't throw. Methods throw. So I would reword it to say "throw from UpdateStatus, WaitForCompletion, ..." or something like that. Having said that. This is also implementation guideline and I was trying to avoid having these as top level bullets in this doc. I would be fine with a section at the end of the doc dedicated to implementing subtypes of Operation<T>

File: docs/dotnet/introduction.md [hunk 11]
```diff
@@ -505,9 +505,20 @@ BlobBaseClient client = ...
...
+
+* If the operation completes with a non-success result, throw ```RequestFailedException``` or its subtype from ```UpdateStatus```, ```WaitForCompletion```, or ```WaitForCompletionAsync```.
+  * Include any relevant error state information in the exception details.
```
Comment (KrzysztofCwalina) [inline docs/dotnet/introduction.md:516]: How? Example code would be needed.

File: docs/dotnet/introduction.md [hunk 12]
```diff
@@ -505,9 +505,20 @@ BlobBaseClient client = ...
...
+* If an underlying service operation call from ```UpdateStatus```, ```WaitForCompletion```, or ```WaitForCompletionAsync``` throws, re-throw ```RequestFailedException``` or its subtype.
+* If the ```Value``` property is evaluated after the operation failed (```HasValue``` is false and ```HasCompleted``` is true) throw the same exception as the one thrown when the operation failed.
```
Comment (KrzysztofCwalina) [inline docs/dotnet/introduction.md:518]: When would it be evaluated? How can the reader know? i.e. code sample would probably clarify that.

**Review by KrzysztofCwalina — CHANGES_REQUESTED** [review]

File: docs/dotnet/introduction.md (hunk 8 above)
Comment (heaths) [reply docs/dotnet/introduction.md:517]: Actually, the `throw` wasn't a perf-related best practice. It's just a best practice from the .NET team even. `throw ex` rewrites the stack, so diagnostics aren't as useful. Just `throw` or - as Chris said - if you're throwing a higher-level exception, wrap the thrown exception as an internal exception. It's stack is retained that way.

File: docs/dotnet/introduction.md [hunk 13]
```diff
@@ -960,6 +951,115 @@ using (JsonDocument json = await JsonDocument.ParseAsync(content, default, cance
 
 Optional JSON properties should be deserialized into nullable model properties.
 
+### Implementing Subtypes of Operation<T> {#dotnet-implement-operation}
+
+{% include requirement/MUST id="dotnet-lro-return" %} check the value of ```HasCompleted``` in subclass implementations of ```UpdateStatus``` and ```UpdateStatusAsync``` and immediately return the result of ```GetRawResponse``` if it is true.
```
Comment (heaths) [inline docs/dotnet/introduction.md:956]: Use a single backtick for inline code. Three backticks are only for code blocks.

File: docs/dotnet/introduction.md [hunk 14]
```diff
@@ -960,6 +951,115 @@ using (JsonDocument json = await JsonDocument.ParseAsync(content, default, cance
...
+  * The exception message should be: "The operation has not yet completed."
+
+```csharp
+// start the operation.
+var operation = await client.StartExampleOperationAsync(someParameter);
+
+if (!operation.HasCompleted)
+{
+    // attempt to get the Value property before the operation has completed.
+    var myResult = operation.Value;  //throws InvalidOperationException.
```
Comment (heaths) [inline docs/dotnet/introduction.md:1054]: No `var` since it shouldn't be in our code anyway (i.e. people won't copy/paste the wrong thing).

**Review by heaths — APPROVED** [review]

File: docs/dotnet/introduction.md [hunk 15]
```diff
@@ -949,6 +951,115 @@ using (JsonDocument json = await JsonDocument.ParseAsync(content, default, cance
...
+```
+
+{% include requirement/MUST id="dotnet-lro-return" %} throw from methods on ```Operation<T>``` subclasses in the following scenarios.
+
+* If an underlying service operation call from ```UpdateStatus```, ```WaitForCompletion```, or ```WaitForCompletionAsync``` throws, re-throw ```RequestFailedException``` or its subtype.
+* If the operation completes with a non-success result, throw ```RequestFailedException``` or its subtype from ```UpdateStatus```, ```WaitForCompletion```, or ```WaitForCompletionAsync```.
+  * Include any relevant error state information in the exception details.
+
+```csharp
+private async ValueTask<Response> UpdateStatusAsync(bool async, CancellationToken cancellationToken)
```
Comment (maririos) [inline docs/dotnet/introduction.md:984]: For people not familiarized with this code I recommend either changing the FR specifics (i.e. `AnalyzeOperationResult` -> `FooResult`) to generic variables and methods, or adding a link to the class that is implementing this so if someone wants more context (like what `ClientCommon           .CreateExceptionForFailedOperationAsync` is, they can take a look)

**Review by maririos — APPROVED** [review]
Thanks Chris for writing this! Supper helpful for future reference

File: docs/dotnet/introduction.md [hunk 16]
```diff
@@ -949,6 +951,115 @@ using (JsonDocument json = await JsonDocument.ParseAsync(content, default, cance
...
+    await UpdateStatusAsync(true, cancellationToken).ConfigureAwait(false);
+
+private async Task<Response> UpdateStatusAsync(bool async, CancellationToken cancellationToken)
+{
+    // Short-circuit when already completed (which improves mocking scenarios that won't have a client).
+    if (HasCompleted)
+    {
+        return GetRawResponse();
+    }
+    ...
```
Comment (KrzysztofCwalina) [inline docs/dotnet/introduction.md:974]: Might be worth showing (or describing) that in else, the implementation would actually do a network call and then return the response. i.e. it's not clear from the guideline that the reason for checking HasCompleted is to avoid hitting the network after the operation is completed.

**Review by KrzysztofCwalina — APPROVED** [review]
